workspaces/orchestrator/plugins/orchestrator/src/api/OrchestratorAuth.ts

```diff
@@ -13,86 +13,151 @@
-
-type ExtractApiRefType<T> = T extends ApiRef<infer U> ? U : never;
-
-export interface OrchestratorAuthDependencies {
-  githubAuthApi: ExtractApiRefType<typeof githubAuthApiRef>;
-  gitlabAuthApi: ExtractApiRefType<typeof gitlabAuthApiRef>;
-  microsoftAuthApi: ExtractApiRefType<typeof microsoftAuthApiRef>;
-}
-export class OrchestratorAuth implements OrchestratorAuthAPi {
-  private readonly deps: OrchestratorAuthDependencies;
-  constructor(deps: OrchestratorAuthDependencies) {
-    this.deps = deps;
-  }
-
-  private async getToken(
-    tokenDescriptor: AuthTokenDescriptor,
-  ): Promise<AuthToken> {
-    let token: string;
-    switch (tokenDescriptor.provider) {
-      case 'github': {
-        if (tokenDescriptor.tokenType === 'openId') {
-          throw new Error(
-            'Github backstage auth API does not support oidc auth tokens',
-          );
-        } else {
-          token = await this.deps.githubAuthApi.getAccessToken(
-            tokenDescriptor.scope,
-          );
+export const useOrchestratorAuth = () => {
+  const app = useApp();
+  const apiHolder = useApiHolder();
+  const githubAuthApi = useApi(githubAuthApiRef);
+  const gitlabAuthApi = useApi(gitlabAuthApiRef);
+  const microsoftAuthApi = useApi(microsoftAuthApiRef);
+
+  const getProviderToken = useCallback(
+    async (
+      authApi: OAuthApi | (OAuthApi & OpenIdConnectApi),
+      tokenDescriptor: AuthTokenDescriptor,
+    ): Promise<string> => {
+      if (tokenDescriptor.tokenType === 'openId') {
+        if (
+          'getIdToken' in authApi &&
+          typeof authApi.getIdToken === 'function'
+        ) {
+          return await authApi.getIdToken();
         }
-        break;
+        throw new Error(
+          `${tokenDescriptor.provider} auth API does not support OpenID Connect tokens`,
+        );
+      } else {
+        return await authApi.getAccessToken(tokenDescriptor.scope);
       }
-      case 'gitlab': {
-        if (tokenDescriptor.tokenType === 'openId') {
-          token = await this.deps.gitlabAuthApi.getIdToken();
-        } else {
-          token = await this.deps.gitlabAuthApi.getAccessToken();
-        }
+    },
+    [],
+  );
 
-        break;
-      }
-      case 'microsoft': {
-        if (tokenDescriptor.tokenType === 'openId') {
-          token = await this.deps.microsoftAuthApi.getIdToken();
-        } else {
-          token = await this.deps.microsoftAuthApi.getAccessToken();
-        }
-        break;
+  const findCustomProvider = useCallback(
+    async (
+      providerApiId: string,
+    ): Promise<OAuthApi | (OAuthApi & OpenIdConnectApi) | undefined> => {
+      const allPlugins = app.getPlugins();
+
+      // Find the API reference for the custom provider
+      const apiRef = allPlugins
+        .flatMap(plugin => Array.from(plugin.getApis()))
+        .filter((api: AnyApiFactory) => api.api.id === providerApiId)
+        .at(0)?.api as ApiRef<OpenIdConnectApi & OAuthApi> | undefined;
```

Comment:
you can simplify this with 
```suggestion
        .find((api: AnyApiFactory) => api.api.id === providerApiId)
        ?.api as ApiRef<OpenIdConnectApi & OAuthApi> | undefined;
```